sled-agent/src/bootstrap/agent.rs

```diff
@@ -729,6 +750,163 @@ impl Agent {
+        sled_hardware::cleanup::delete_omicron_vnics(&self.log)
+            .await
+            .map_err(BootstrapError::Cleanup)?;
+        illumos_utils::opte::delete_all_xde_devices(&self.log)?;
+        Ok(())
+    }
+
+    async fn uninstall_storage_locked(
+        &self,
+        _state: &tokio::sync::MutexGuard<'_, SledAgentState>,
+    ) -> Result<(), BootstrapError> {
+        let datasets = zfs::get_all_omicron_datasets()
+            .map_err(BootstrapError::ZfsDatasetsList)?;
+        for dataset in &datasets {
+            info!(self.log, "Removing dataset: {dataset}");
+            zfs::Zfs::destroy_dataset(dataset)?;
+        }
+
+        Ok(())
+    }
+
+    /// Resets this sled, removing:
+    ///
+    /// - All control plane zones (except the switch zone)
+    /// - All sled-local configuration
+    /// - All underlay networking
+    /// - All storage managed by the control plane
+    ///
+    /// This API is intended to put the sled into a state where it can
+    /// subsequently be initialized via RSS.
+    pub async fn sled_reset(&self) -> Result<(), BootstrapError> {
+        let mut state = self.sled_state.lock().await;
+
+        if let SledAgentState::After(_) = &mut *state {
+            // We'd like to stop the old sled agent before starting a new
+            // hardware monitor -- however, if we cannot start a new hardware
+            // monitor, the bootstrap agent may be in a degraded state.
+            let server = match std::mem::replace(
+                &mut *state,
+                SledAgentState::Before(None),
+            ) {
+                SledAgentState::After(server) => server,
+                _ => panic!(
+                    "Unexpected state (we should have just matched on it)"
+                ),
+            };
+            server.close().await.map_err(BootstrapError::SledError)?;
+        };
+
+        // Try to reset the sled, but do not exit early on error.
+        let result = async {
+            self.uninstall_zones_locked(&state).await?;
+            self.uninstall_sled_local_config_locked(&state).await?;
+            self.uninstall_networking_locked(&state).await?;
+            self.uninstall_storage_locked(&state).await?;
+            Ok::<(), BootstrapError>(())
+        }
+        .await;
+
+        // Try to restart the bootstrap agent hardware monitor before
+        // returning any errors from reset.
+        match &mut *state {
+            SledAgentState::Before(None) => {
+                let hardware_monitor = self.start_hardware_monitor()
+                    .await
+                    .map_err(|err| {
+                        warn!(self.log, "Failed to restart bootstrap agent hardware monitor");
+                        err
+                    })?;
+                *state = SledAgentState::Before(Some(hardware_monitor));
+            }
+            _ => {}
+        };
+
+        // Return any errors encountered resetting the sled.
+        result?;
```

Comment:
Trivial nitpick - it looks like these types match and we could just end the function with

```suggestion
        result
```

Reply:
Updated